ow_plexil/src/plexil-adapter/PlexilPlanSelection.cpp

```diff
@@ -0,0 +1,132 @@
+    plan_array.push_back(initial_plan);
+  }
+  //workaround for the getPlanState not returning correctly for first plan
+  m_first_plan = true;
+  
+  //initialize subscriber and publisher
+  m_planSelectionCommandSubscriber = std::make_unique<ros::Subscriber>
+      (m_genericNodeHandle->
+       subscribe("/plexil_plan_selection_commands", 20,
+                 &PlexilPlanSelection::planSelectionCommandsCallback, this));
+  m_planSelectionStatusPublisher = std::make_unique<ros::Publisher>
+      (m_genericNodeHandle->advertise<std_msgs::String>
+       ("/plexil_plan_selection_status", 20));
+
+  ROS_INFO("Plexil node started, ready for PLEXIL plans.");
+}
+
+
+void PlexilPlanSelection::start()
+{
+  ros::Rate rate(10); // 10 Hz for overall rate we are spinning
+  ros::Rate rate2(1); // 1 Hz for timeout sleep
+  std_msgs::String status;
+  while(ros::ok()){
+    //if we have plans in our plan array we begin the control loop
+    if(plan_array.size() > 0){
+      int length = plan_array.size();
+      for(int i = 0; i < length; i++){
+        //try to run the plan
+        if(m_executive->runPlan(plan_array[0].c_str())){
+          //workaround for getPlanState not working on first plan
+          if(m_first_plan == true){
+            m_first_plan = false; 
+          }
+          int timeout = 0;
+          // Times out after 3 seconds or the plan is registered as running.
+          while(m_executive->getPlanState() && timeout < 30){
+            ros::spinOnce();
+            rate.sleep();
+            timeout+=1;
+            if(timeout % 10 == 0){
+              ROS_ERROR("Plan not responding, timing out in %i seconds", (3 - timeout/10));
+            }
+          }
+            //if timed out we set plan as failed for GUI
+            if(timeout == 30){
+              ROS_INFO ("Plan timed out, try again.");
+              status.data = "FAILED:" + plan_array[0];
+              m_planSelectionStatusPublisher->publish(status);
+            }
+            //otherwise we set it as running
+            else{
+                status.data = "SUCCESS:" + plan_array[0];
+                m_planSelectionStatusPublisher->publish(status);
+            }
+        }
+        //if error from run() we set as failed for GUI
+        else{
+            status.data = "FAILED:" + plan_array[0];
+            m_planSelectionStatusPublisher->publish(status);
+        }
+        //delete the plan we just ran from plan array
+        plan_array.erase(plan_array.begin());
+        //wait for current plan to finish before running next plan
+        while(!m_executive->getPlanState() && m_first_plan == false){
+          ros::spinOnce();
+          rate.sleep();
+        }
+        //Once plan is finished set status to complete for GUI
+        status.data = "COMPLETE";
+        m_planSelectionStatusPublisher->publish(status);
+        //checks if plan_array has been cleared
+        if(plan_array.size() == 0){
+          break; 
+        }
+      }
+    }
```

Comment:
This is somewhat complicated code and seems hacky. I recommend it to be split and simplified.